be/src/formats/parquet/level_builder.cpp

```diff
@@ -0,0 +1,493 @@
+        int64_t milliseconds = data_col[idx].to_unix_second() * 1000;
+        values.push_back(milliseconds);
+    });
+
+    write_leaf_callback(LevelBuilderResult{
+            .num_levels = ctx.size(),
+            .def_levels = def_levels->data(),
+            .rep_levels = rep_levels->data(),
+            .values = reinterpret_cast<uint8_t*>(values.data()),
+            .null_bitset = nullptr,
+    });
+
+    return Status::OK();
+}
+
+Status LevelBuilder::_write_varchar_column_chunk(const LevelBuilderContext& ctx, const TypeDescriptor& type_desc,
+                                                 const ::parquet::schema::NodePtr& node, const ColumnPtr& col,
+                                                 const CallbackFunction& write_leaf_callback) {
+    auto data_col = down_cast<const RunTimeColumnType<TYPE_VARCHAR>*>(ColumnHelper::get_data_column(col.get()));
+    const auto null_col = get_raw_null_column(col);
+    auto& vo = data_col->get_offset();
+    auto& vb = data_col->get_bytes();
+
+    // Use the rep_levels in the context from caller since node is primitive.
+    auto rep_levels = ctx._rep_levels;
+    auto def_levels = _make_def_levels(ctx, node, null_col);
+
+    std::vector<::parquet::ByteArray> values;
+    values.reserve(ctx.size());
+    do_ignore_null(ctx, null_col, [&](int idx) {
+        auto len = static_cast<uint32_t>(vo[idx + 1] - vo[idx]);
+        auto ptr = reinterpret_cast<const uint8_t*>(vb.data() + vo[idx]);
+        values.emplace_back(len, ptr);
+    });
+
+    write_leaf_callback(LevelBuilderResult{
+            .num_levels = ctx.size(),
+            .def_levels = def_levels->data(),
+            .rep_levels = rep_levels->data(),
+            .values = reinterpret_cast<uint8_t*>(values.data()),
+            .null_bitset = nullptr,
+    });
+    return Status::OK();
+}
+
+Status LevelBuilder::_write_array_column_chunk(const LevelBuilderContext& ctx, const TypeDescriptor& type_desc,
+                                               const ::parquet::schema::NodePtr& node, const ColumnPtr& col,
+                                               const CallbackFunction& write_leaf_callback) {
+    DCHECK(type_desc.type == TYPE_ARRAY);
+    auto outer_node = std::static_pointer_cast<::parquet::schema::GroupNode>(node);
+    auto mid_node = std::static_pointer_cast<::parquet::schema::GroupNode>(outer_node->field(0));
+    auto inner_node = mid_node->field(0);
+
+    const auto null_col = get_raw_null_column(col);
+    const auto array_col = down_cast<ArrayColumn*>(ColumnHelper::get_data_column(col.get()));
+    const auto& elements = array_col->elements_column();
+    const auto& offsets = array_col->offsets_column()->get_data();
+    LevelBuilderContext derived_ctx(ctx._max_def_level + node->is_optional() + 1, ctx._max_rep_level + 1,
+                                    ctx.size() + elements->size());
+
+    int subcol_size = 0;
+    for (auto i = 0; i < ctx.size(); i++) {
```

Comment:
you could memset def level and rep level at beginning, then modify them for null and empty case. it will be more cache friendly and less branch predication